### PR TITLE
Support for reference convert_fx working on gpu

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61403

Summary:
This PR enables gpu only quantization, best used with is_reference since
there are not many gpu kernels for ops as of now.

This PR mainly changes how qconfigs and their obs constructors operate once the qconfig they are in
get put onto a specific module. The function add_module_to_qconfig_obs_ct takes the obs constructors on the original
qconfig, and configures them so that when invoked, the created obs will
be on whatever device the module occupies. (Once observers are created,
module.to(device) is already setup so that it moves any observers). To do this,
a new method and a few small chanegs were added to the _PartialWrapper class that
our observers already use to create constructors (without changing the
existing functionality). These changes work in
concert with changes to the prepare flow such that when the qconfigs are
propagated to the moduels (in quantize.py and qconfig_utils.py) they are configured using add_module_to_qconfig_obs_ctr.

Ideally this would work on other models but the is_reference support for
a lot of modules isn't there yet, those tests should be added in a
future PR

Test Plan: python test/test_quantization.py
TestQuantizeFxModels.test_static_gpu_convert_basic

python test/test_quantization.py
TestQuantizeFxModels.test_switch_device_prepare_convert

python test/test_quantization.py TestQuantizeFxModels.test_prepare_serialize_switch_device_convert

python test/test_quantization.py TestQuantizeFx.test_qconfig_precedence

Reviewers:

Subscribers:

Tasks:

Tags:

still not working

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

mend

changed method so the constructor is interactable

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

just saving since server seems to restart randomly

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29684114](https://our.internmc.facebook.com/intern/diff/D29684114)

### CODE REVIEW

File: torch/nn/quantized/modules/utils.py [hunk 1]
```diff
@@ -4,7 +4,9 @@
 from torch.nn.modules.module import _addindent
 
 def _quantize_weight(float_wt, observer):
-    wt_scale, wt_zp = observer.calculate_qparams()
+    device = float_wt.device
+    wt_scale, wt_zp = observer.calculate_qparams()  # may be better to make calculate qparams give on correct device
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/utils.py:8]: can we make calculate_qparams to return scale/zero_point in correct device?

Comment (jerryzh168) [reply torch/nn/quantized/modules/utils.py:8]: yeah, agreed with the comment

File: torch/quantization/observer.py [hunk 2]
```diff
@@ -1106,7 +1112,7 @@ def calculate_qparams(self):
                 "must run observer before calling calculate_qparams.\
                                     Returning default scale and zero point "
             )
-            return torch.tensor([1.0]), torch.tensor([0])
+            return torch.tensor([1.0]).to('cuda'), torch.tensor([0]).to('cuda')
```
Comment (raghuramank100) [inline torch/quantization/observer.py:1115]: Can we move the tensors to self.device?

Comment (HDCharles) [reply torch/quantization/observer.py]: changed in more recent version

File: torch/nn/quantized/modules/utils.py (hunk 1 above)
Comment (HDCharles) [reply torch/nn/quantized/modules/utils.py]: this change is in the most recent version

File: torch/quantization/fx/qconfig_utils.py [hunk 3]
```diff
@@ -13,6 +13,21 @@
 QConfigAny = Union[torch.quantization.QConfig,
                    torch.quantization.QConfigDynamic, None]
 
+def add_device_to_obs_ctr_in_qconfig(qconfig_dict, device='cpu'):
+    if '' in qconfig_dict.keys() and isinstance(qconfig_dict[''], torch.quantization.QConfig):
+        factory_kwargs = {'device': device}
+        new_activation = qconfig_dict[''].activation.with_args(factory_kwargs=factory_kwargs)
```
Comment (raghuramank100) [inline torch/quantization/fx/qconfig_utils.py:19]: Where is the device attr used in the observer?

File: torch/quantization/fx/qconfig_utils.py [hunk 4]
```diff
@@ -13,6 +13,21 @@
 QConfigAny = Union[torch.quantization.QConfig,
                    torch.quantization.QConfigDynamic, None]
 
+def add_device_to_obs_ctr_in_qconfig(qconfig_dict, device='cpu'):
+    if '' in qconfig_dict.keys() and isinstance(qconfig_dict[''], torch.quantization.QConfig):
```
Comment (supriyar) [inline torch/quantization/fx/qconfig_utils.py:17]: Why is this only specific to keys that are `''`. For other entries in the qconfig dict I think we need to have similar init?

File: torch/quantization/fx/qconfig_utils.py (hunk 3 above)
Comment (supriyar) [reply torch/quantization/fx/qconfig_utils.py:19]: Also this assumes all observers will pass in the factory_kwargs to the buffers that it creates. Would this break if the user has their own observer code or we have buffers that don't use the kwargs?

File: torch/nn/quantized/modules/conv.py [hunk 5]
```diff
@@ -205,8 +207,8 @@ def get_qconv(cls, mod, activation_post_process, weight_post_process=None):
                     mod.stride, mod.padding, mod.dilation, mod.groups,
                     mod.bias is not None, mod.padding_mode)
         qconv.set_weight_bias(qweight, mod.bias)
-        qconv.scale = float(act_scale)
-        qconv.zero_point = int(act_zp)
+        qconv.scale = act_scale.to(torch.float)
+        qconv.zero_point = act_zp.to(torch.long)
```
Comment (supriyar) [inline torch/nn/quantized/modules/conv.py:211]: The scale/zero_point are scalar attributes in the module. After these changes will they be tensors?

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

File: torch/quantization/fx/qconfig_utils.py (hunk 3 above)
Comment (HDCharles) [reply torch/quantization/fx/qconfig_utils.py:19]: it wouldn't break, it just wouldn't actually do anything in those cases.All that would happen is that if the user doesn't set up their custom observer correctly, it just wouldn't work with gpu quantization.

It seems fairly low risk since that argument seems entirely unused as is. factory_kwargs is specifically intended to handle device info so anything that utilizes it would already need to be able to handle a device argument.

File: torch/quantization/fx/qconfig_utils.py (hunk 4 above)
Comment (HDCharles) [reply torch/quantization/fx/qconfig_utils.py:17]: this is fixed in the most recent version. Although I'm not sure if there's a better way to handle other types, it seems like other qconfig types don't necessarily live in qconfig_dict["<key_name>"].

File: torch/nn/quantized/modules/conv.py (hunk 5 above)
Comment (HDCharles) [reply torch/nn/quantized/modules/conv.py:211]: I removed this change per your comments

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

File: test/quantization/fx/test_quantize_fx.py [hunk 6]
```diff
@@ -1159,7 +1159,7 @@ def forward(self, x):
             "module_name_regex": [("module_conv*", module_name_regex_qconfig)],
             "module_name": [("module_conv2", module_name_qconfig)]}
         m = prepare_fx(m, qconfig_dict)
-        self.assertEqual(m.linear.qconfig, global_qconfig)
+        self.assertEqual(m.linear.qconfig, qconfig_dict[""])
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:1162]: should we revert this change?

File: torch/quantization/fx/qconfig_utils.py [hunk 7]
```diff
@@ -13,6 +13,22 @@
 QConfigAny = Union[torch.quantization.QConfig,
                    torch.quantization.QConfigDynamic, None]
 
+def add_device_to_obs_ctr_in_qconfig(qconfig_dict, device='cpu'):
```
Comment (jerryzh168) [inline torch/quantization/fx/qconfig_utils.py:16]: do we want to modify qconfig_dict? I feel it would be better to keep it since it's user input?

Also I feel it might be better to provide this argument when we actually insert observers instead of changing the configs, wondering why this does not work? https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/prepare.py#L219

File: torch/quantization/fx/utils.py [hunk 8]
```diff
@@ -360,7 +360,8 @@ def create_getattr_from_value(module: torch.nn.Module, graph: Graph, prefix: str
     """
     get_new_attr_name = get_new_attr_name_with_prefix(prefix)
     attr_name = get_new_attr_name(module)
-    module.register_buffer(attr_name, torch.tensor(value))
+    device = assert_and_get_unique_device(module)
+    module.register_buffer(attr_name, torch.tensor(value).to(device))
```
Comment (jerryzh168) [inline torch/quantization/fx/utils.py:364]: nit: torch.tensor(value, device=device) would be faster I think since it creates a tensor directly on the device

Comment (jerryzh168) [reply torch/quantization/fx/utils.py:364]: same for other cases

File: test/quantization/fx/test_quantize_fx.py [hunk 9]
```diff
@@ -4735,3 +4760,7 @@ def test_resnet18_ddp(self):
         model = models.__dict__[name](pretrained=True).eval().float()
         self._test_model_impl(
             'ddp', 'resnet18', model, eager_quantizable_model)
+
+if __name__ == "__main__":
+    a = TestQuantizeFxModels()
+    a.test_static_gpu_convert_mobilenet()
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:4766]: can you change this to https://github.com/pytorch/pytorch/blob/master/test/quantization/eager/test_quantize_eager_ptq.py#L1255, we'd like to run tests with python test/test_quantization.py TestCalss.test_case so that we don't need to worry about files (one less thing to remember)

File: torch/quantization/fx/qconfig_utils.py (hunk 7 above)
Comment (HDCharles) [reply torch/quantization/fx/qconfig_utils.py:16]: The issue is that for weights and other observers that don't actually get inserted, they exist only as a constructor until they are instantiated and used. (https://fburl.com/diffusion/1k1fdoea). So the options are to edit the constructor or edit the from_float function for every module. There's probably a way to edit only the constructor that gets propagated to the modules, but when I tried that (https://fburl.com/diffusion/mt1l207v) , it seems to insert the constructor at multiple different points and it overwrites it.

File: torch/quantization/fx/utils.py (hunk 8 above)
Comment (HDCharles) [reply torch/quantization/fx/utils.py:364]: ok, will fix

File: test/quantization/fx/test_quantize_fx.py (hunk 9 above)
Comment (HDCharles) [reply test/quantization/fx/test_quantize_fx.py:4766]: oh ya that should have been removed.

Comment (HDCharles) [reply test/quantization/fx/test_quantize_fx.py]: (fixed)

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

File: test/quantization/fx/test_quantize_fx.py (hunk 6 above)
Comment (HDCharles) [reply test/quantization/fx/test_quantize_fx.py]: yeah reverted

File: torch/quantization/fx/qconfig_utils.py (hunk 7 above)
Comment (HDCharles) [reply torch/quantization/fx/qconfig_utils.py]: I'm going to remove these changes, it seems like there's significant disagreement about the desired direction.

File: torch/quantization/fx/qconfig_utils.py (hunk 3 above)
Comment (HDCharles) [reply torch/quantization/fx/qconfig_utils.py]: I'm removing these changes in any case, it seems unclear as to how the team wants to solve this issue and its not actually a blocker for this PR

File: torch/quantization/fx/utils.py [hunk 10]
```diff
@@ -360,7 +360,8 @@ def create_getattr_from_value(module: torch.nn.Module, graph: Graph, prefix: str
     """
     get_new_attr_name = get_new_attr_name_with_prefix(prefix)
     attr_name = get_new_attr_name(module)
-    module.register_buffer(attr_name, torch.tensor(value))
+    device = assert_and_get_unique_device(module)
```
Comment (raghuramank100) [inline torch/quantization/fx/utils.py:363]: Is this solution specific to fx? If we want this approach to work for eager, would we need to modify the from_float method?

File: test/quantization/fx/test_quantize_fx.py [hunk 11]
```diff
@@ -4735,3 +4760,8 @@ def test_resnet18_ddp(self):
         model = models.__dict__[name](pretrained=True).eval().float()
         self._test_model_impl(
             'ddp', 'resnet18', model, eager_quantizable_model)
+
+if __name__ == '__main__':
```
Comment (raghuramank100) [inline test/quantization/fx/test_quantize_fx.py:4764]: Good to see this error message being added

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

File: torch/quantization/fx/utils.py (hunk 10 above)
Comment (HDCharles) [reply torch/quantization/fx/utils.py:363]: All of the changes I made affect the prepare code at the point where it has called the non-fx prepare code so it should be applicable to both. There may be a slight differece but it should be easy to make work on eager, at least I cannot see a reason why it wouldn't.

File: torch/quantization/fx/qconfig_utils.py (hunk 7 above)
Comment (HDCharles) [reply torch/quantization/fx/qconfig_utils.py]: edit: changes were made such that the original qconfig is unaffected

File: test/quantization/fx/test_quantize_fx.py (hunk 6 above)
Comment (HDCharles) [reply test/quantization/fx/test_quantize_fx.py]: had to do some significant finagling to get this test working, It works correctly but the assertEqual kept breaking due to how the m.linear.qconfig would get updated with the new constructor device info. Its working now though.

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

File: torch/quantization/fx/convert.py [hunk 12]
```diff
@@ -158,7 +158,9 @@ def convert(model: GraphModule, is_reference: bool = False,
     qconfig_map: Dict[str, QConfigAny] = model._qconfig_map  # type: ignore[assignment]
 
     # move to cpu since we only have quantized cpu kernels
-    model.eval().cpu()
+    model.eval()
+    if not is_reference:
+        model.cpu()
```
Comment (vkuzo) [inline torch/quantization/fx/convert.py:163]: I would recommend removing this altogether, I don't think quantization code should change the device, this is a job of the user.  Changing it here is unexpected.

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

File: torch/quantization/qconfig.py [hunk 13]
```diff
@@ -138,3 +138,36 @@ def assert_valid_qconfig(qconfig: Optional[Union[QConfig, QConfigDynamic]],
         )
         assert not is_per_channel, \
             'Per channel weight observer is not supported yet for ConvTranspose{n}d.'
+
+QConfigAny = Union[QConfig, QConfigDynamic, None]
+
+# should probably be in qconfig_utils but get circular reference issues TODO
+def add_device_to_obs_ctr_in_qconfig(qconfig: QConfigAny, module: torch.nn.Module):
```
Comment (vkuzo) [inline torch/quantization/qconfig.py:145]: This unfortunately does couple device with `qconfig`, because we are encoding `QConfig` with the device and saving it in a object of type `QConfig`.  Could we estimate the engineering time to keep device out of qconfig and specify it directly when objects are constructed, so it's consistent with how other things work in PyTorch?

If we do really need this approach, could we move this logic out of anything related to `QConfig`, and name the resulting object differently (`QConfigAndDevice`, etc) to emphasize that this is no longer just quantization configs, there is an additional parameter here unrelated to quantization

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

File: torch/quantization/qconfig.py (hunk 13 above)
Comment (HDCharles) [reply torch/quantization/qconfig.py:145]: how about the new version that only ties it to module? is that philosophically at least an acceptable direction? Implementation aside?

File: torch/quantization/fx/convert.py (hunk 12 above)
Comment (HDCharles) [reply torch/quantization/fx/convert.py:163]: Ok, will do

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

File: torch/quantization/qconfig.py [hunk 14]
```diff
@@ -122,7 +122,61 @@ def get_default_qat_qconfig(backend='fbgemm'):
...
+    """
+    def __new__(cls, activation=torch.nn.Identity, weight=torch.nn.Identity):
+        # catch common mistakes
+        if isinstance(weight, nn.Module):
+            raise ValueError("QConfigWithModule received observer instance, please pass observer class instead. " +
+                             "Use MyObserver.with_args(x=1) to override arguments to constructor if needed")
+        return super(QConfigWithModule, cls).__new__(cls, activation, weight)
+
+
+def create_qconfig_with_module(
```
Comment (vkuzo) [inline torch/quantization/qconfig.py:153]: could we check if this is going to survive serialization?  For example, at any time the user should be able to save their model, load it from disk, and everything should be the same.

File: torch/quantization/qconfig.py [hunk 15]
```diff
@@ -122,7 +122,61 @@ def get_default_qat_qconfig(backend='fbgemm'):
         qconfig = default_qat_qconfig
     return qconfig
 
-def assert_valid_qconfig(qconfig: Optional[Union[QConfig, QConfigDynamic]],
+class QConfigWithModule(namedtuple('QConfigWithModule', ['activation', 'weight'])):
```
[the remark this answers is not in the document]
Comment (HDCharles) [reply torch/quantization/qconfig.py:125]: yeah, thats easy enough to change.

File: torch/quantization/observer.py [hunk 16]
```diff
@@ -42,6 +53,27 @@ def _with_args(cls_or_self, **kwargs):
     r = _PartialWrapper(partial(cls_or_self, **kwargs))
     return r
 
+def _with_callable_args(cls_or_self, **kwargs):
```
[the remark this answers is not in the document]
Comment (HDCharles) [reply torch/quantization/observer.py:56]: I can just wrap function definitions and pass those. The problem is that then you can't interact with them. You can't add arguments...etc. 

It'd look something like v14in phabricator: https://fburl.com/diff/arxgfzeh

File: torch/quantization/qconfig.py (hunk 14 above)
Comment (HDCharles) [reply torch/quantization/qconfig.py:153]: TestPostTrainingDynamic.test_quantized_rnn and python test/test_quantization.py TestObserver.test_per_channel_observers both pass, (as do TestPostTrainingDynamic and TestObserver).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

File: torch/quantization/qconfig.py (hunk 14 above)
Comment (HDCharles) [reply torch/quantization/qconfig.py]: I also added a new test case, seems like no issue though it looks like qconfig isn't serialized, and since already created observers were mostly functioning correctly, I don't think this change affects serialization as far as I can tell.

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

File: torch/quantization/fx/convert.py (hunk 12 above)
Comment (HDCharles) [reply torch/quantization/fx/convert.py:166]: actually, I can't remove this, looks like some processes like qat are assuming everything coming in is on cpu. Added a todo instead, looks like maybe this should be removed once qat is supported on gpu

File: torch/quantization/observer.py [hunk 17]
```diff
@@ -273,7 +307,7 @@ def _calculate_qparams(
             zero_points: Zero points tensor of shape (#channels,)
         """
         if not check_min_max_valid(min_val, max_val):
-            return torch.tensor([1.0]), torch.tensor([0])
+            return torch.tensor([1.0]).to(min_val.device.type), torch.tensor([0]).to(min_val.device.type)
```
Comment (vkuzo) [inline torch/quantization/observer.py:310]: would it work to specify the device at initialization, like `torch.tensor([1.0], device=min_val.device)`?

File: torch/quantization/qconfig.py [hunk 18]
```diff
@@ -138,3 +138,52 @@ def assert_valid_qconfig(qconfig: Optional[Union[QConfig, QConfigDynamic]],
         )
         assert not is_per_channel, \
             'Per channel weight observer is not supported yet for ConvTranspose{n}d.'
+
+QConfigAny = Union[QConfig,
```
Comment (vkuzo) [inline torch/quantization/qconfig.py:142]: this exists in `torch.quantization.fx.qconfig_utils`, could we move it out of fx folder and reuse it instead of creating a new one?

**Review by vkuzo — APPROVED** [review]
looks good, had some minor comments inline

File: torch/quantization/observer.py (hunk 17 above)
Comment (HDCharles) [reply torch/quantization/observer.py:310]: yeah that would be better

File: torch/quantization/qconfig.py (hunk 18 above)
Comment (HDCharles) [reply torch/quantization/qconfig.py:142]: yeah, will do

File: torch/quantization/fx/convert.py (hunk 12 above)
Comment (raghuramank100) [reply torch/quantization/fx/convert.py:166]: QAT should be supported on GPU. Even without GPU inference, this is needed.  I will create an issue to track this support.

File: torch/quantization/observer.py [hunk 19]
```diff
@@ -1106,7 +1140,7 @@ def calculate_qparams(self):
                 "must run observer before calling calculate_qparams.\
                                     Returning default scale and zero point "
             )
-            return torch.tensor([1.0]), torch.tensor([0])
+            return torch.tensor([1.0]).to(self.min_val.device.type), torch.tensor([0]).to(self.min_val.device.type)
```
Comment (raghuramank100) [inline torch/quantization/observer.py:1143]: Similar comment to the previous one from Vasiliy here

**Review by raghuramank100 — COMMENTED** [review]
Looks great, thanks for the detailed documentation

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29684114).

**Comment by HDCharles** [comment]
this was meged
